addons/website_sale/tests/test_website_sale_cart.py

```
@@ -355,6 +355,36 @@ def test_remove_archived_product_line(self):
             # Assert: the line has been removed
             self.assertFalse(order.order_line)
 
+    def test_product_not_allowed_in_cart_checkout(self):
+        """If an order has a line containing a product that is not allowed in the cart,
+        it should raise an user error during checkout."""
+        # Arrange
+        user = self.user_portal
+        website = self.website.with_user(user)
+        user.write(self.dummy_partner_address_values)
+        website = self.website.with_user(user)
+        product = self.env['product.product'].create({
+            'name': 'Product_test',
+            'sale_ok': True,
+            'website_published': True,
+        })
+        with MockRequest(self.env(user=user), website=website):
+            self.WebsiteSaleController.cart_update_json(product_id=product.id, add_qty=1)
+            order = website.sale_get_order()
+            order.partner_id = user.partner_id.id
+            # pre-condition: the order contains an active product
+            self.assertRecordValues(order.order_line, [{
+                "product_id": product.id,
+            }])
+            self.assertTrue(product.active)
+            # Act: archive the product and open the cart
+            product.active = False
+            with self.assertRaises(UserError):
```

Make an assert about the content of the error message, here is an example:
https://github.com/odoo/odoo/blob/f41b36e2b0125b4f8b6a3e36ee4f5ccac8871360/addons/project/tests/test_project_sharing.py#L431
And to make it clear that you only want to raise an error in the last action perform the other calls outside of the `with`